### ISSUE TITLE
Revamping cranelift_wasm's `{Func,Module}Environment` and public-facing API

### ISSUE THREAD

**Issue by fitzgen** [CraneStation/cranelift#1080]
Since I'm still fairly new to the code base, and since this ended up a bit more involved than I expected it to be, I figured it would be good to bounce these ideas off y'all before whipping up an implementation and submitting a PR.

### Problem

With the introduction of [support for multi-value Wasm](https://github.com/CraneStation/cranelift/pull/1049), we need to maintain some state about the whole module's translation<sup id="back-foot-0">[0](#foot-0)</sup> that we later use during each function's translation. However, `cranelift`'s `{Module,Func}Environment` traits and `wasmtime`'s implementation of these traits doesn't make it easy to get that shared module translation state when we need it to translate a particular function.

Here is some psuedo code describing the current state of `wasmtime`'s interactions with `cranlift_wasm` (my understanding is that SpiderMonkey takes a similar approach). The nesting roughly represents the call stack.

```
let data: &'data [u8] = ...;
let module_env: impl cranelift_wasm::ModuleEnvironment<'data> = ...;
let module_translation: wasmtime::ModuleTranslation = module_env.translate();
    cranelift_wasm::translate_module()
        enqueues function bodies into module_env's inner wasmtime::ModuleTranslation
parallel for each function body in module_translation:
    let func_env: impl cranelift_wasm::FuncEnvironment = ...;
    cranelift_wasm::FuncTranslator::translate(&mut func_env)
```

The shared module translation state needs to be initialized while we are decoding the Wasm module inside `cranelift_wasm::translate_module`, but it needs to be used during `cranelift_wasm::FuncTranslator::translate`.

### Desiridata

1. We want the creation, upkeep, and lifetime of the shared module translation state to be managed by `cranelift_wasm` and not external API consumers. Since the module translation state is the same for every user of `cranelift_wasm`, we don't want to put its interface into the `*Environment` traits, and force every `cranelift_wasm` API consumer to implement the same set of methods the same way. That's just busy work.

2. We would like to use plain borrows to access the shared module translation state during function translation. The module translation state is relatively temporary: it is only required during function translation, and afterwards we can forget about it. It is additionally only needed in a read-only fashion. This suggests allocating the module translation state on the stack and borrowing it during function translation. However, this is a bit tricky because `cranelift_wasm::translate_module` is not an ancestor stack frame above `cranelift_wasm::FuncTranslate::translate`.

3. We would like it to be difficult -- impossible even -- to use the shared module state for module A while translating a function from module B. At worst, if this mixup occurred, there would be a verification error, so this point is more of a nice-to-have than a show stopper. (This is of course assuming that we always verify the clif in, for example, `wasmtime-jit`. If we ever call JIT'd functions without verifying the clif, then this point is a must-have for memory safety!)

### Proposed Solution

* Introduce `ModuleTranslationState` for the Wasm module's associated translation state that is shared across all functions' translations. This is analagous to the current `TranslationState` but associated with a module rather than a function.

* We would rename to `FuncTranslationState` to help differentiate between module's and function's associated states.

* `ModuleEnvironment<'data>` would become `ModuleEnvironment<'state>` where `'state` is a `ModuleTranslationState`'s lifetime. Similarly, `FuncEnvironment` would become `FuncEnvironment<'state>`. We would thread a `&ModuleTranslationState<'state>` parameter through all these trait methods.

  For example:

  ```rust
  trait ModuleEnvironment<'state> {
      // ...

      fn define_function_body(
          &mut self,
          state: &ModuleTranslationState<'state>,
          body_bytes: &'state [u8],
          body_offset: usize,
      ) -> WasmResult<()>;

      // ...
  }
  ```

* We would provide a "scoped" API that "brands" its state and data to a single, invariant lifetime. When combined with replacing `'data` with `'state`, this gets the borrow checker to guarantee (3) for us, because the lifetime can't be extended, contracted, or unified with any other liftime by rustc.

  ```rust
  impl ModuleTranslationState {
      pub fn scoped<F, T>(data: &[u8], f: F) -> T
      where
          F: for<'state> FnOnce(&mut ModuleTranslationState<'state>) -> T
      {
          let mut state = ModuleTranslationState::new(data);
          f(&mut state)
      }

      /// Get the underlying data with our unique, invariant lifetime.
      pub fn data(&self) -> &'state [u8] {
          self.data
      }
  }
  ```

* Because ownership of the state stays inside `scoped`, API consumers never have to manage it themselves. While references to it are passed to API consumers, they just pass it along to other `cranelift_wasm` API functions. This gets us (1).

Here is the pseudocode for what `wasmtime`'s new usage of `cranelift_wasm` would look like:

```
let data: &'data [u8] = ...;
cranelift_wasm::ModuleTranslationState::scoped(data, |state: &ModuleTranslationState<'state>| {
    let module_env: impl cranelift_wasm::ModuleEnvironment<'state> = ...;
    let module_translation: wasmtime::ModuleTranslation = module_env.translate();
        cranelift_wasm::translate_module()
            enqueues function bodies into module_env's inner wasmtime::ModuleTranslation
    parallel for each function body in module_translation:
        let func_env: impl cranelift_wasm::FuncEnvironment<'state> = ...;
        cranelift_wasm::FuncTranslator::translate(&mut func_env)
});
```

#### Further Improvements

From here, there are some orthogonal API clean ups we could make, namely converting `cranelift_wasm::translate_module` and `cranelift_wasm::FuncTranslator::translate` into methods of `ModuleTranslationState`:

```
impl<'state> ModuleTranslationState<'state> {
    /// Note: `&mut self` since this populates the shared module state.
    pub fn translate_module(
        &mut self,
        module_env: &mut dyn ModuleEnvironment<'state>,
    ) -> WasmResult<()> {
        // ...
    }

    /// Note: `&self` since this only reads the already-populated state. This
    /// allows it to be shared across threads when translating functions in
    /// parallel.
    pub fn translate_function(
        &self,
        func_env: &mut dyn FuncEnvironment<'state>,
    ) -> WasmResult<()> {
        // ...
    }
}
```

This would make it pretty hard to even get (3)-style compilation errors by accident, and would further improve the API's usage ergonomics:

```
let data: &'data [u8] = ...;
cranelift_wasm::ModuleTranslationState::scoped(data, |state: &ModuleTranslationState<'state>| {
    let module_env: impl cranelift_wasm::ModuleEnvironment<'state> = ...;
    state.translate_module(&mut env);
        enqueues function bodies into module_env's inner wasmtime::ModuleTranslation

    let module_translation: wasmtime::ModuleTranslation = module_env.result;
    parallel for each function body in module_translation:
        let func_env: impl cranelift_wasm::FuncEnvironment<'state> = ...;
        state.translate_function(&mut func_env);
});
```

### Alternatives

* If we drop the nice-to-have (3), we can avoid the word-y `scoped` API, avoid introducing closures and indentation, and avoid threading invariant lifetimes through a bunch of functions, types, and traits. Instead, we would return the `ModuleTranslationState` from `cranelift::translate_module`, save it in `wasmtime::ModuleTranslation`, and then finally pass shared references to it back into each call of `cranelift_wasm::FuncTranslator::translate`. This does, however, make the user manage the module translation state's lifetime themselves, as well.

  If we pursued this further, we could still make `cranelift_wasm::FuncTranslator::translate` into a method of `ModuleTranslationState`, which would improve the ergonomics of function translation, IMO.

  The API usage would then end up something like this:

  ```
  let data: &'data [u8] = ...;
  let module_env: impl cranelift_wasm::ModuleEnvironment<'data> = ...;
  let state: ModuleTranslationState<'data> = cranelift_wasm::translate_module(&mut env);
      enqueues function bodies into module_env's inner wasmtime::ModuleTranslation

  let module_translation: wasmtime::ModuleTranslation = module_env.result;
  parallel for each function body in module_translation:
      let func_env: impl cranelift_wasm::FuncEnvironment = ...;
      state.translate_function(&mut func_env);
  ```

  I think this is also pretty good, and ultimately the decision comes down to how much we care about statically enforcing (3).

----------

<sup id="foot-0">0</sup> To be precise, we need to keep the Wasm module's decoded type section around so we can determine the type signature for a given multi-value block later on when translating a function. [↩](#back-foot-0)

**Comment by fitzgen** [comment]
+cc  @sunfishcode, @abrown, @bnjbvr, @sstangl, @yurydelendik

Interested in your thoughts on this. Thanks!

**Comment by pchickey** [comment]
This proposal looks great to me! Thanks for working on this.

**Comment by bnjbvr** [comment]
Thanks for this proposal, it makes a lot of sense. For what it's worth, Spidermonkey doesn't use the ModuleTranslation API at all: we do our own validation first, which collects the equivalent of ModuleTranslationState (it is called ModuleEnvironment there), and the trait methods on FuncEnvironment we implement in Spidermonkey query information from this shared state (e.g. during `make_global`) and cache them locally, leading to duplicated state. Of course it's safe because all this state is immutable as you note, so there's no risk of caches getting out of sync. Having a ModuleTranslationState could help making some of these FuncEnvironment methods optional, when no particular treatment is needed (Spidermonkey has special behavior for each of these methods, so wouldn't benefit from it). Moving state from the TranslationState to the ModuleTranslationState sounds good, since it would avoid copies.

> Since the module translation state is the same for every user of cranelift_wasm, we don't want to put its interface into the *Environment traits, and force every cranelift_wasm API consumer to implement the same set of methods the same way.

Slightly unsure about this, because this disallows experimenting with new proposals that aren't implemented in Cranelift (e.g. think about proposals updating semantics, like the mutable globals proposal did). That could be worked around by having ModuleTranslationState be a trait with default implementations of trait methods, though.

## Regarding requirement (3), i.e. not passing one module's state to another module's functions translation

I understand it's nice to have such a requirement in that it avoids making issues, but is this an issue that could possibly happen in practice? i.e., does Wasmtime compile sometimes functions from different modules within the body of a single function? Are there cases where multiple binary wasm modules are passed as part of data? If the answer is no to both of these questions, is this something that Wasmtime is likely to implement in a near future? Spidermonkey compiles only one module at once, and I don't expect this to change, so the issue is nonexistent for this use case. It's nice to have the property that each module is compiled independently from others, even if they're compiled in parallel. I am really asking because the `scoped` method seems to introduce some complexity, while putting state on the stack and passing references to it would catch the same borrowing issues *as long as there's only one module compiled at a time*.

## One note about the alternative

> Instead, we would return the ModuleTranslationState from cranelift::translate_module, save it in wasmtime::ModuleTranslation, and then finally pass shared references to it back into each call of cranelift_wasm::FuncTranslator::translate

Could instead the `FuncTranslator` itself keep a reference (annotated with lifetime) to the `ModuleTranslationState`, passed in the ctor or in a `reset` function? (so as to reuse memory allocations from one function translation to the next)

## Further away improvements: validate wasm as we translate it in Cranelift

If we have at our disposal the module translation state, we could even validate wasm code as we translate it from Cranelift. This is something embedders have to do right now before translating. Some of the information required is already here, in the duplicated state stored in TranslationState, but as we implement more proposals, we might need even more of it (e.g. bulk memory requires knowledge about elem segments). If I understand correctly, this proposal would help getting rid of these state copies entirely as a first step. The next logical step would then be to move validation within Cranelift itself.

Spidermonkey has some special usage for a subset of the ModuleTranslationState, used for caching wasm modules and being able to re-instantiate them later (e.g. initial elem / data segments that are used during instantiation). That's not a problem though, since this could be extracted from the ModuleTranslationState once translating the module is done. (It would involve some interesting work, for instance to make streaming compilation work, but we're not there yet.)

**Comment by fitzgen** [comment]
@bnjbvr, thanks for the detailed response!

> Moving state from the TranslationState to the ModuleTranslationState sounds good, since it would avoid copies.

To be clear, everything I'm proposing is just for new stuff introduced when adding support for multi-value Wasm. AFAIK, there isn't anything being copied between `TranslationState`s right now. If, however, my understanding is incorrect, and there is stuff being copied for each `TranslationState` that we don't mutate, then it would make sense to de-dupe it here.

> Slightly unsure about this, because this disallows experimenting with new proposals that aren't implemented in Cranelift (e.g. think about proposals updating semantics, like the mutable globals proposal did). That could be worked around by having ModuleTranslationState be a trait with default implementations of trait methods, though.

For functions, we have the `FuncEnvironment` trait for embedders to customize function-level translation, and the `TranslationState` "internal", non-trait, non-customizable structure.

We have a `ModuleEnvironment` trait where embedders can customize module-level translation right now. But there is currently no `ModuleTranslationState` "internal", non-trait, non-customizable structure.

Part of what I'm proposing is extending this dual environment/state, customizable/internal pairs to modules, not just functions. Assuming we agree those duals makes sense, the question comes down to "what should be customizable (via living in the `*Environment` trait) and what should not (via living in the `*State` structure)?"

For wasm multi-value, the shared module state that each function needs access to is the decoded types section and the map from a type index to the params and returns. I don't see much to customize here, since the only non-side-effectful thing an embedder could do differently is return the wrong type.

Agreed that it makes sense to put things on the `*Environment` traits for when we are customizing the actual translation to cranelift's IR and choosing instructions (as opposed to looking up data that we already parsed and stored on the side).

> does Wasmtime compile sometimes functions from different modules within the body of a single function? Are there cases where multiple binary wasm modules are passed as part of data? If the answer is no to both of these questions, is this something that Wasmtime is likely to implement in a near future? Spidermonkey compiles only one module at once, and I don't expect this to change, so the issue is nonexistent for this use case.

It doesn't do these things right now. I could imagine maaayyyybe wanting to do it in some future world where "shared nothing linking with wasm interface types" is a thing, but that's it.

> I am really asking because the scoped method seems to introduce some complexity, while putting state on the stack and passing references to it would catch the same borrowing issues as long as there's only one module compiled at a time.

Yeah, *if the cranelift IR is always verified before calling any JIT'd functions* then it likely isn't worth going full `scoped` and invariant lifetime.

However, if the cranelift IR is not verified before calling JIT'd functions, then I think the case for actually getting the borrow checker to statically guarantee that we are using the right module state is much stronger.

> Could instead the `FuncTranslator` itself keep a reference (annotated with lifetime) to the `ModuleTranslationState`, passed in the ctor or in a `reset` function? (so as to reuse memory allocations from one function translation to the next)

Yeah, the `FuncTranslator` pretty much wraps a `TranslationState` so either that or passing it into `FuncTranslator::translate` should work.

> Further away improvements: validate wasm as we translate it in Cranelift

Yeah definitely. Yury's `wasmparser` crate does support validating-while-parsing, so you know that everything you're given is valid Wasm.

What I'm more concerned with in (3) is that we are then transforming that valid Wasm into valid cranelift IR.

Great point about bulk-memory -- that is another instance of where we can be given valid Wasm, but then later when we translate that valid Wasm to cranelift IR, we need to be sure we aren't mixing up indices or misinterpreting things that we parsed earlier, etc.

**Comment by bnjbvr** [comment]
> If, however, my understanding is incorrect, and there is stuff being copied for each TranslationState that we don't mutate, then it would make sense to de-dupe it here.

I did look at the implementation in Spidermonkey, so maybe it's different for other projects and YMMV, but, to give a concrete example, let's consider globals. A global can be used by several functions, and is supposed to be stored at the same exact location (its location shouldn't depend on the current function that we're compiling). So while there's the need for a user-facing API to control how to retrieve the global's content location, the resulting `ir::Global` should be the same for all the functions in a single module. This data is now created per each function, and stored in each `TranslationState`. This might be orthogonal, though.

> I don't see much to customize here, since the only non-side-effectful thing an embedder could do differently is return the wrong type.

You could imagine the embedder would want to restrict acceptable types to a subset of types, too. Multi-value has some ABI changes implications, so Spidermonkey could use such a way to filter observed types when decoding wasm. I'm mostly playing the devil's advocate here.

**Comment by fitzgen** [comment]
> Multi-value has some ABI changes implications, so Spidermonkey could use such a way to filter observed types when decoding wasm.

This is quite a good point, in my opinion.

**Comment by fitzgen** [comment]
PR for this (not the invariant lifetime version) over in https://github.com/CraneStation/cranelift/pull/1111
